### ISSUE TITLE
internal error while loading sample parquet file

### ISSUE THREAD

**Issue by jbellis** [kaskada-ai/kaskada#202]
Following the instructions in getting started with jupyter, on a local install:

![image](https://user-images.githubusercontent.com/42158/229601281-ec59bd67-2c35-4a39-82c6-2e3205da347a.png)

Environment is python 3.10.6 on Ubuntu 22.04

**Comment by therapon** [comment]
I did try this on my Ubuntu 22.04 and it did work. 

1. What is the version of jupyter you are running `jupyter --version` on your command line should print out the version, here is mine 
```
> jupyter --version 
Selected Jupyter core packages...
IPython          : 7.34.0
ipykernel        : 6.21.2
ipywidgets       : 8.0.2
jupyter_client   : 8.0.2
jupyter_core     : 5.2.0
jupyter_server   : 2.3.0
jupyterlab       : 3.6.1
nbclient         : 0.7.2
nbconvert        : 7.2.9
nbformat         : 5.7.3
notebook         : 6.5.2
qtconsole        : 5.3.2
traitlets        : 5.9.0

```

2. The logs for the engine and manager should be under `~/.cache/kaskada` 
For each instantiation there should be a folder under this path whose name is a UUID. Hopefully there is only 1 folder :) else a `grep` on the request id that shoes in the error message would tell you which file it is. 

If you can have a look at the logs and paste here any errors that would be helpful.

**Comment by jbellis** [comment]
```
$ cat ~/.cache/kaskada/logs/a0414e39-1d8e-4412-838f-88fd7b9da963/engine_stderr.txt 
/home/jonathan/.cache/kaskada/bin/engine@v0.5.4/kaskada-engine: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
```

I manually installed libssl1.1 as suggested here https://stackoverflow.com/questions/72133316/libssl-so-1-1-cannot-open-shared-object-file-no-such-file-or-directory and that fixed it.  still, I'm curious

1. did you also get libssl1.1 on ubuntu 22.04, or is your installation not looking for it?
2. what is it that requires libssl1.1?  it looks like ubuntu 22.04 has deprecated that in favor of libssl3, which is why it requires manual installation

here is my jupyter version output
```
$ jupyter --version
Selected Jupyter core packages...
IPython          : 7.34.0
ipykernel        : 6.22.0
ipywidgets       : 8.0.6
jupyter_client   : 8.1.0
jupyter_core     : 5.3.0
jupyter_server   : 2.5.0
jupyterlab       : not installed
nbclient         : 0.7.2
nbconvert        : 7.2.10
nbformat         : 5.8.0
notebook         : 6.5.3
qtconsole        : 5.4.1
traitlets        : 5.9.0
```

**Comment by therapon** [comment]
> ```
> $ cat ~/.cache/kaskada/logs/a0414e39-1d8e-4412-838f-88fd7b9da963/engine_stderr.txt 
> /home/jonathan/.cache/kaskada/bin/engine@v0.5.4/kaskada-engine: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
> ```

Ah yes. Ubuntu 22.xx moved to libssl 3.x while ubuntu 20.xx is on libssl 1.x

> I manually installed libssl1.1 as suggested here https://stackoverflow.com/questions/72133316/libssl-so-1-1-cannot-open-shared-object-file-no-such-file-or-directory and that fixed it. still, I'm curious
> 
>     1. did you also get libssl1.1 on ubuntu 22.04, or is your installation not looking for it?

Yes I did. A while ago so it slipped my mind. 

 
>     2. what is it that requires libssl1.1?  it looks like ubuntu 22.04 has deprecated that in favor of libssl3, which is why it requires manual installation
> 

Yes that is why you need to manually install. 
Last I checked it was one of our dependencies that requires is that is unfortunately non-Rust code (might be rocksdb but not sure). But I will double check and report back here.

**Comment by therapon** [comment]
On who brings ssl dynamically the answer is `openssl-sys` crate 

```
> cargo clean && cargo build -vv 2>/dev/null | grep 'rustc-link-lib' | grep -v static 
[openssl-sys 0.9.83] cargo:rustc-link-lib=ssl
[openssl-sys 0.9.83] cargo:rustc-link-lib=crypto
[librocksdb-sys 0.8.3+7.4.4] cargo:rustc-link-lib=stdc++
```
The first 2 lines are the relevant ones. These are the ones that add `-lssl` and `-lcrypto` to the linking stage of the sparrow binary resulting in a dynamic dependency to libssl. 

It looks like our dependency on `pulsar` is bringing `openssl-sys` in. (from running `cargo tree`) 

Here is the relevant part of the tree 


```
│   │   ├── pulsar v5.1.0
│   │   │   ├── async-native-tls v0.4.0
│   │   │   │   ├── futures-util v0.3.27 (*)
│   │   │   │   ├── native-tls v0.2.11
│   │   │   │   │   ├── log v0.4.17 (*)
│   │   │   │   │   ├── openssl v0.10.48
│   │   │   │   │   │   ├── bitflags v1.3.2
│   │   │   │   │   │   ├── cfg-if v1.0.0
│   │   │   │   │   │   ├── foreign-types v0.3.2
│   │   │   │   │   │   │   └── foreign-types-shared v0.1.1
│   │   │   │   │   │   ├── libc v0.2.140
│   │   │   │   │   │   ├── once_cell v1.17.1
│   │   │   │   │   │   ├── openssl-macros v0.1.0 (proc-macro)
│   │   │   │   │   │   │   ├── proc-macro2 v1.0.54 (*)
│   │   │   │   │   │   │   ├── quote v1.0.26 (*)
│   │   │   │   │   │   │   └── syn v1.0.109 (*)
│   │   │   │   │   │   └── openssl-sys v0.9.83
│   │   │   │   │   │       └── libc v0.2.140
│   │   │   │   │   │       [build-dependencies]
│   │   │   │   │   │       ├── autocfg v1.1.0
│   │   │   │   │   │       ├── cc v1.0.79 (*)
│   │   │   │   │   │       └── pkg-config v0.3.26
│   │   │   │   │   ├── openssl-probe v0.1.5
│   │   │   │   │   └── openssl-sys v0.9.83 (*)

```

It looks like the `openssl-sys` has a new release (as of 2 days ago) that supports libssl 3.x https://github.com/sfackler/rust-openssl/pull/1264

**Comment by jnorrid** [comment]
Same issue on macOS 13.3 on an M1 but using the CLI version of the Hello World walk-through located here:
https://kaskada.io/docs-site/kaskada/main/getting-started/hello-world-cli.html and specifically, at this step: https://kaskada.io/docs-site/kaskada/main/getting-started/hello-world-cli.html#loading-data...

./kaskada-cli load \
>     --table Purchase \
>     --file-path file://purchase.parquet
2:17PM INF starting load
2:17PM WRN no client-id found, initiating request without passing client-id header.
2:17PM FTL  error="rpc error: code = Internal desc = internal error"

It is likely the same libssl1.1 dependency...

**Comment by jnorrid** [comment]
Hi @therapon - Thank you for the response :) The kaskada-engine starts just fine, see below. That was never an issue. The failure I am picking up is when I use kaskada-cli to try and connect and execute the load: kaskada-cli --> kaskada-manager --> kaskada-engine.

```
536 jnorrid@tadakatsu:kaskada$ ./kaskada-engine serve
2023-04-04T22:40:42.588983Z  WARN Sparrow in batch-mode{build_info=BuildInfo { sparrow_version: "0.5.0", github_ref: "engine@v0.5.4", github_sha: "39f8043ff27983cce5bf824e76102eeb40dc221b", github_workflow: "Engine Release" }}:load_region{provider=ImdsRegionProvider { client: LazyClient { client: OnceCell { value: None }, builder: Builder { max_attempts: None, endpoint: None, mode_override: None, token_ttl: None, connect_timeout: None, read_timeout: None, config: Some(ProviderConfig { env: Env(Real), fs: Fs(Real), sleep: Some(TokioSleep), region: None }) } }, env: Env(Real) }}: failed to load region from IMDS err=Failed to load session token: timeout: error trying to connect: HTTP connect timeout occurred after 1s
2023-04-04T22:40:43.593100Z  WARN Sparrow in batch-mode{build_info=BuildInfo { sparrow_version: "0.5.0", github_ref: "engine@v0.5.4", github_sha: "39f8043ff27983cce5bf824e76102eeb40dc221b", github_workflow: "Engine Release" }}:load_region{provider=ImdsRegionProvider { client: LazyClient { client: OnceCell { value: None }, builder: Builder { max_attempts: None, endpoint: None, mode_override: None, token_ttl: None, connect_timeout: None, read_timeout: None, config: Some(ProviderConfig { env: Env(Real), fs: Fs(Real), sleep: Some(TokioSleep), region: None }) } }, env: Env(Real) }}: failed to load region from IMDS err=Failed to load session token: timeout: error trying to connect: HTTP connect timeout occurred after 1s
2023-04-04T22:40:43.595700Z  INFO Sparrow in batch-mode{build_info=BuildInfo { sparrow_version: "0.5.0", github_ref: "engine@v0.5.4", github_sha: "39f8043ff27983cce5bf824e76102eeb40dc221b", github_workflow: "Engine Release" }}: S3 Config: Config
2023-04-04T22:40:43.600821Z  INFO Sparrow in batch-mode{build_info=BuildInfo { sparrow_version: "0.5.0", github_ref: "engine@v0.5.4", github_sha: "39f8043ff27983cce5bf824e76102eeb40dc221b", github_workflow: "Engine Release" }}: Starting Sparrow listening on 0.0.0.0:50052
```

Adding ${PWD} did end up working for me, so it looks like I had a different problem, although I am receiving a warning regarding a missing client-id: 

```
./kaskada-cli load \
    --table Purchase \
    --file-path file://${PWD}/purchase.parquet
```
```
539 jnorrid@tadakatsu:kaskada$ ./kaskada-cli sync apply -f spec.yaml
5:45PM INF starting apply
5:45PM WRN no client-id found, initiating request without passing client-id header.
5:45PM INF resource not found on system, will create it kind=*kaskadav1alpha.Table name=Purchase
5:45PM INF created resource with provided spec kind=*kaskadav1alpha.Table name=Purchase
5:45PM INF Success!
```

**Comment by therapon** [comment]
Glad to see that you could get the command to succeed. 

I see that warning as well on my end. A quick look around the code it seems that the client-id is used as part of metadata that gets logged. So while we should probably add some form of client-id to be passed by the CLI, it is not affecting the behaviour of the system to the best of my knowledge. 

@rltvty  maybe you have more context about the warning and the use of `client-id`. Here is the line that I think is logging the warning: https://github.com/kaskada-ai/kaskada/blob/79b2d94eefdd0337218473a67bd9bd319ad03abe/clients/cli/api/api.go#LL189C2-L189C2

**Comment by rltvty** [comment]
`client-id` exists in the system to provide multi-tenacy when using Kaskada in a production environment.  Each client essentially gets their own namespace for tables, views, queries, etc... 

When working locally in your own instance of Kaskada, this isn't a concern.  I think we should probably just change this log line to `debug` level.  I'll make a PR.

**Comment by kerinin** [comment]
Also, to be clear there are two distinct issues here, and we should fix both of them:
1. We should try to accept inputs in more cases
2. We should return an actionable error if we can't use an input

**Comment by therapon** [comment]
> This issue has also come up on every user study I've conducted

Do you mean the libssl version issue or the error message text we use for errors (which seems to be internal error most of the times with little further information)?

**Comment by kerinin** [comment]
Doh - yeah I meant to post that https://github.com/kaskada-ai/kaskada/discussions/208

**Comment by epinzur** [comment]
this should be fixed by the file existence check added on #301
